File: tasks/python3/mergeConfigWithWMTS.py
#!/usr/bin/env python

from util import dict_merge
from copy import deepcopy
from optparse import OptionParser
import os
import sys
import json

prog = os.path.basename(__file__)
base_dir = os.path.join(os.path.dirname(__file__), "..")
version = "2.0.0"
help_description = """\
Concatenates all configuration items a directory into one configuration file.
"""

def wmts_dict_merge(target, *args):
  # Merge multiple dicts
  # Recursively merge dicts and set non-dict values
  obj = args[0]['layers']
  conf = args[1]['layers']

  if not isinstance(obj, dict):
    return obj
  for k, v in obj.items():
    if k in conf and isinstance(conf[k], dict):
      foundSourceMisMatch = False
      if 'projections' in v and 'projections' in conf[k]:
        conf_projections = conf[k]['projections']
        for projectionKey, projection in v['projections'].items():
          source = projection['source']
          if projectionKey in conf_projections:
            if 'source' in conf_projections[projectionKey]:
              if source != conf_projections[projectionKey]['source']:
                foundSourceMisMatch = True
                continue
      if foundSourceMisMatch:
        continue
      if k in target and isinstance(target[k], dict):
          dict_merge(target[k], v, conf[k])
      else:
          target[k] = dict_merge(v, conf[k])
  return target

# MAIN
parser = OptionParser(usage="Usage: %s <input_dir> <output_file>" % prog,
                      version="%s version %s" % (prog, version),
                      epilog=help_description)

(options, args) = parser.parse_args()
if len(args) != 2:
    parser.error("Invalid number of arguments")
input_dir = args[0]
output_file = args[1]
new_conf = {}
new_conf["layers"] = {}
new_conf["sources"] = {}

with open(output_file, "r", encoding="utf-8") as fp:
  output_data = json.load(fp)
file_count = 0
for file in os.listdir(input_dir):
  try:
      if not file.endswith(".json"):
        continue
      file_count += 1
      with open(os.path.join(input_dir, file), "r", encoding="utf-8") as fp:
          data = json.load(fp)
      new_conf['layers'] = wmts_dict_merge(new_conf['layers'], data, output_data)
      new_conf["sources"] = dict_merge(new_conf["sources"], data['sources'], output_data['sources'])
  except Exception as e:
      sys.stderr.write("ERROR: %s: %s\n" %
                        (os.path.join(input_dir, file), str(e)))
      sys.exit(1)
new_conf = dict_merge(new_conf, output_data)
json_options = {}
json_options["indent"] = 2
json_options["separators"] = (',', ': ')

with open(output_file, "w", encoding="utf-8") as fp:
    json.dump(new_conf, fp, **json_options)

print("%s: %s file(s) merged into %s" % (prog, file_count,
    os.path.basename(output_file)))

File: tasks/build-options/mergeConfigWithWMTS.js
const fs = require('fs')
const path = require('path')
const yargs = require('yargs')
const { dictMerge } = require('./util')
const prog = path.basename(__filename)

const options = yargs
  .usage('Usage: $0 [inputDir] [outputDir]')
  .option('inputDir', {
    demandOption: true,
    alias: 'i',
    type: 'string',
    description: 'getcapabilities input directory'
  })
  .option('outputFile', {
    demandOption: true,
    alias: 'o',
    type: 'string',
    description: 'wmts output directory'
  })
  .epilog('Concatenates all configuration items a directory into one configuration file.')

const { argv } = options
if (!argv.inputDir && !argv.outputFile) {
  throw new Error('Invalid number of arguments')
}

const inputDir = argv.inputDir
const outputFile = argv.outputFile

let newConf = {}
let layers = {}
newConf.layers = {}
newConf.sources = {}

const outputData = JSON.parse(fs.readFileSync(outputFile, 'utf-8'))
let fileCount = 0

async function main () {
  layers = newConf.layers
  delete layers.sources
  delete newConf.layers
  newConf = dictMerge(newConf, outputData)
  await processFiles()

  const jsonOptions = {}
  jsonOptions.indent = 2
  jsonOptions.separators = [',', ': ']

  await fs.writeFileSync(outputFile, JSON.stringify(newConf, jsonOptions))

  console.warn(`${prog}: ${fileCount} file(s) merged into ${path.basename(outputFile)}`)
}

async function processFiles () {
  try {
    const files = fs.readdirSync(inputDir)
    for (const file of files) {
      if (!file.endsWith('.json')) continue
      fileCount += 1
      try {
        const data = fs.readFileSync(path.join(inputDir, file), 'utf-8')
        const jsonData = JSON.parse(data)
        newConf.layers = await wmtsDictMerge(layers, jsonData, outputData)
        newConf.sources = dictMerge(newConf.sources, jsonData.sources, outputData.sources)
      } catch (error) {
        throw new Error(`ERROR: ${path.join(inputDir, file)}: ${error.message}`)
      }
    }
  } catch (error) {
    console.error(error)
  }
}

async function wmtsDictMerge (target, obj, conf) {
  if (!(obj instanceof Object)) {
    return obj
  }
  for (const [k, v] of Object.entries(obj)) {
    if (k in conf && typeof conf[k] === 'object') {
      let foundSourceMisMatch = false
      if ('projections' in v && 'projections' in conf[k]) {
        const confProjections = conf[k].projections
        for (const [projectionKey, projection] of Object.entries(v.projections)) {
          const source = projection.source
          if (projectionKey in confProjections) {
            if ('source' in confProjections[projectionKey]) {
              if (source !== confProjections[projectionKey].source) {
                foundSourceMisMatch = true
                continue
              }
            }
          }
        }
      }
      if (foundSourceMisMatch) {
        continue
      }
      if (k in target && typeof target[k] === 'object') {
        target[k] = dictMerge(target[k], v, conf[k])
      } else {
        target[k] = dictMerge(v, conf[k])
      }
    }
  }
  return target.layers
}

main().catch((err) => {
  console.error(err.stack)
})
